x-pack/metricbeat/module/openai/usage/usage.go

```diff
@@ -0,0 +1,392 @@
+//
+// For each API key:
+// 1. Generates a secure SHA-256 hash of the key for state tracking
+// 2. Checks the state store for the last processed date
+// 3. Adjusts start date if previous state exists to avoid duplicate collection
+// 4. Iterates through each day in the range, collecting usage data
+// 5. Updates state store with the latest processed date
+func (m *MetricSet) fetchDateRange(startDate, endDate time.Time, httpClient *RLHTTPClient) error {
+	for _, apiKey := range m.config.APIKeys {
+		// SHA-256 is a cryptographic hash function that generates a 256-bit (32-byte) digest,
+		// represented as a 64-character hexadecimal string. The hash function is deterministic,
+		// ensuring the same input consistently produces identical output, while being computationally
+		// infeasible to reverse. Its strong collision resistance makes it highly suitable for
+		// secure key storage.
+		//
+		// The hexadecimal representation uses only alphanumeric characters [0-9a-f], making it
+		// ideal for cross-platform filename compatibility. The fixed-length output provides
+		// predictable storage requirements and consistent behavior across different systems.
+		hasher := sha256.New()
+		hasher.Write([]byte(apiKey.Key))
+		hashedKey := hex.EncodeToString(hasher.Sum(nil))
+		stateKey := "state_" + hashedKey
+
+		if m.stateStore.Has(stateKey) {
+			lastProcessedDate, err := m.stateStore.Get(stateKey)
+			if err != nil {
+				m.logger.Errorf("Error reading state for API key: %v", err)
+				continue
+			}
+
+			lastDate, err := time.Parse("2006-01-02", lastProcessedDate)
+			if err != nil {
+				m.logger.Errorf("Error parsing last processed date: %v", err)
+				continue
+			}
+
+			startDate = lastDate.AddDate(0, 0, 1)
+			if startDate.After(endDate) {
+				continue
+			}
+		}
+
+		for d := startDate; !d.After(endDate); d = d.AddDate(0, 0, 1) {
+			dateStr := d.Format("2006-01-02")
+			if err := m.fetchSingleDay(dateStr, apiKey.Key, httpClient); err != nil {
+				m.logger.Errorf("Error fetching data for date %s: %v", dateStr, err)
+				continue
+			}
+		}
+
+		if err := m.stateStore.Put(stateKey, endDate.Format("2006-01-02")); err != nil {
+			m.logger.Errorf("Error storing state for API key: %v", err)
+		}
+	}
+	return nil
+}
+
+func (m *MetricSet) fetchSingleDay(dateStr, apiKey string, httpClient *RLHTTPClient) error {
+	req, err := m.createRequest(dateStr, apiKey)
+	if err != nil {
+		return fmt.Errorf("error creating request: %w", err)
+	}
+
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("error making request: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("error response from API: %s", resp.Status)
+	}
+
+	return m.processResponse(resp, dateStr)
+}
+
+func (m *MetricSet) createRequest(dateStr, apiKey string) (*http.Request, error) {
+	req, err := http.NewRequestWithContext(context.TODO(), http.MethodGet, m.config.APIURL, nil)
```

Comment:
if there is no context available, why not use just `NewRequest`?

Reply:
Because of linter. `noctx` checker will cry otherwise. Anyways, http.NewRequest also calls `NewRequestWithContext` internally and uses context.Background.

So didn't change this.